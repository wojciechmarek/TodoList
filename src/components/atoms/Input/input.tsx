import { useCallback, useEffect, useState } from 'react';
import { InputProps } from './input.interface';
import { InputStyle } from './input.styled';

export const Input = (props: InputProps) => {
  const { value, placeholder, handleInputChange, ...params } = props;

  const [phrase, setPhrase] = useState('');

  const onChange = useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
    const { value } = e.target;
    handleInputChange(value);
    setPhrase(value);
  }, []);

  useEffect(() => {
    setPhrase(value);
  }, [value]);

  return (
    <InputStyle
      value={phrase}
      onChange={onChange}
      placeholder={placeholder}
      {...params}
    />
  );
};
