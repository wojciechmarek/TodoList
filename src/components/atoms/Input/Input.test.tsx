import { configure, render } from '@testing-library/react';
import { Input } from './input';
import { InputProps } from './input.interface';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import TestRenderer from 'react-test-renderer';

describe('Input', () => {
  let inputProps: InputProps;

  beforeEach(() => {
    configure({ throwSuggestions: true });

    inputProps = {
      value: 'hello world',
      placeholder: 'placeholder test',
      handleInputChange: jest.fn(),
    };
  });

  it('should render', () => {
    const { baseElement } = render(<Input {...inputProps} />);
    expect(baseElement).toBeTruthy();
  });

  it('should match the snapshot', () => {
    const tree = TestRenderer.create(<Input {...inputProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('when the input is rendered', () => {
    it('should have the correct value', () => {
      const { getByRole } = render(<Input {...inputProps} />);
      const input = getByRole('textbox');
      expect(input).toHaveValue('hello world');
    });

    it('should have the correct placeholder', () => {
      const { getByRole } = render(<Input {...inputProps} />);
      const input = getByRole('textbox');
      expect(input).toHaveAttribute('placeholder', 'placeholder test');
    });
  });

  describe('when the text to the input is entered', () => {
    it('should call the handleInputChange function when the input changes', async () => {
      const { getByRole } = render(<Input {...inputProps} />);
      const input = getByRole('textbox');
      await userEvent.type(input, 'qwerty', { delay: 1 });
      expect(inputProps.handleInputChange).toHaveBeenCalled();
    });

    it('should call the handleInputChange function 3 times for 3 characters', async () => {
      const { getByRole } = render(<Input {...inputProps} />);
      const input = getByRole('textbox');
      await userEvent.type(input, 'abc', { delay: 1 });
      expect(inputProps.handleInputChange).toHaveBeenCalledTimes(3);
    });

    it('should call the handleInputChange function with correct phrase', async () => {
      const { getByRole } = render(<Input {...inputProps} />);
      const input = getByRole('textbox');
      await userEvent.type(input, '123', { delay: 1 });
      expect(inputProps.handleInputChange).toHaveBeenCalledWith(
        'hello world123'
      );
    });
  });
});
