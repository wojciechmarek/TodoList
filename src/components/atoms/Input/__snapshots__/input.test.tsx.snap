// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should match the snapshot 1`] = `
.emotion-0 {
  border: 1px solid var(--search-input-border-and-icon-bg-color);
  background-color: var(--search-input-bg-color);
  color: var(--input-font-color);
  font-weight: 800;
  padding-left: 1em;
  border-radius: var(--standard-border-radius);
  -webkit-transition: filter 0.3s ease-in-out;
  transition: filter 0.3s ease-in-out;
}

.emotion-0::-webkit-input-placeholder {
  color: var(--text-font-color);
  font-weight: 600;
  opacity: 0.4;
}

.emotion-0::-moz-placeholder {
  color: var(--text-font-color);
  font-weight: 600;
  opacity: 0.4;
}

.emotion-0:-ms-input-placeholder {
  color: var(--text-font-color);
  font-weight: 600;
  opacity: 0.4;
}

.emotion-0::placeholder {
  color: var(--text-font-color);
  font-weight: 600;
  opacity: 0.4;
}

.emotion-0:hover {
  -webkit-filter: brightness(0.9);
  filter: brightness(0.9);
}

<input
  className="emotion-0"
  onChange={[Function]}
  placeholder="placeholder test"
  value=""
/>
`;
